modules/c++/re/unittests/test_regex.cpp

```diff
@@ -120,20 +202,75 @@ TEST_CASE(testMultilineBehavior)
 
 TEST_CASE(testSub)
 {
-    re::RegexMatch matches;
-    re::Regex rx("ar");
-    std::string subst = rx.sub("Hearo", "ll");
+    // Part of the intent here is to make sure we can handle strings
+    // substituted that are longer or shorter than what they're
+    // replacing
+    re::Regex rx("arb");
+    std::string subst = rx.sub("Hearbo", "ll");
     TEST_ASSERT_EQ(subst, "Hello");
-    subst = rx.sub("Hearo Keary!", "ll");
+
+    subst = rx.sub("Hearbo Kearby!", "ll");
     TEST_ASSERT_EQ(subst, "Hello Kelly!");
+
+    subst = rx.sub("Hearbo Kearby!", "llll");
+    TEST_ASSERT_EQ(subst, "Hellllo Kelllly!");
+
+    // So what happens if we're replacing a pattern with something that would
+    // also match that pattern itself?
+    subst = rx.sub("Hearbo Kearby!", "arbarb");
+    TEST_ASSERT_EQ(subst, "Hearbarbo Kearbarby!");
+
+    // And now replace it with nothing at all
+    subst = rx.sub("Hearbo Kearby!", "");
+    TEST_ASSERT_EQ(subst, "Heo Key!");
+
+    // Make sure we can replace stuff right next to each other
+    subst = rx.sub("arbarbarb!", "blah");
+    TEST_ASSERT_EQ(subst, "blahblahblah!");
 }
 
 TEST_CASE(testSplit)
 {
-    re::RegexMatch matches;
-    re::Regex rx("ar");
+    re::Regex rx1("ar");
     std::vector<std::string> vec;
-    rx.split("ONEarTWOarTHREE", vec);
+    rx1.split("ONEarTWOarTHREE", vec);
+    TEST_ASSERT_EQ(vec.size(), 3);
+    TEST_ASSERT_EQ(vec[0], "ONE");
+    TEST_ASSERT_EQ(vec[1], "TWO");
+    TEST_ASSERT_EQ(vec[2], "THREE");
+
+    vec.clear();
+    rx1.split("ONEarTWOarTHREEar", vec);
+    TEST_ASSERT_EQ(vec.size(), 3);
+    TEST_ASSERT_EQ(vec[0], "ONE");
+    TEST_ASSERT_EQ(vec[1], "TWO");
+    TEST_ASSERT_EQ(vec[2], "THREE");
+
+    re::Regex rx2("x");
+    vec.clear();
+    rx2.split("ONExTWOxTHREE", vec);
+    TEST_ASSERT_EQ(vec.size(), 3);
+    TEST_ASSERT_EQ(vec[0], "ONE");
+    TEST_ASSERT_EQ(vec[1], "TWO");
+    TEST_ASSERT_EQ(vec[2], "THREE");
+
+    vec.clear();
+    rx2.split("ONExTWOxTHREEx", vec);
+    TEST_ASSERT_EQ(vec.size(), 3);
+    TEST_ASSERT_EQ(vec[0], "ONE");
+    TEST_ASSERT_EQ(vec[1], "TWO");
+    TEST_ASSERT_EQ(vec[2], "THREE");
+
+    re::Regex rx3("xxxxxxxxxx");
+    vec.clear();
+    rx3.split("ONExxxxxxxxxxTWOxxxxxxxxxxTHREE", vec);
+    TEST_ASSERT_EQ(vec.size(), 3);
+    TEST_ASSERT_EQ(vec[0], "ONE");
+    TEST_ASSERT_EQ(vec[1], "TWO");
+    TEST_ASSERT_EQ(vec[2], "THREE");
+
+    vec.clear();
+    rx3.split("ONExxxxxxxxxxTWOxxxxxxxxxxTHREExxxxxxxxxx", vec);
```

Comment:
So I noticed in split(), if the match is at the beginning, we return an empty string in the vector, but if the match is at the end, we don't.  I feel like these should behave the same way (I know in Python it includes the empty string for both).